File: dataloom-backend/app/api/endpoints/projects.py
"""Project CRUD API endpoints.

Handles upload, retrieval, save (checkpoint), and revert operations.
"""

import uuid
from pathlib import Path

from fastapi import APIRouter, Depends, File, Form, HTTPException, UploadFile
from fastapi.responses import FileResponse
from sqlalchemy.exc import SQLAlchemyError
from sqlmodel import Session

from app import database, models, schemas
from app.api.dependencies import get_current_user, get_project_or_404
from app.services.file_service import delete_project_files, get_original_path, store_upload
from app.services.project_service import (
    create_checkpoint,
    create_project,
    delete_change_log,
    delete_project,
    get_last_change_log,
    get_recent_projects,
)
from app.services.transformation_service import apply_logged_transformation
from app.utils.file_formats import get_format
from app.utils.logging import get_logger
from app.utils.pandas_helpers import dataframe_to_response, read_table_safe, save_table_safe
from app.utils.security import validate_upload_file

logger = get_logger(__name__)

router = APIRouter()


@router.post("/upload", response_model=schemas.ProjectResponse)
async def upload_project(
    file: UploadFile = File(...),
    projectName: str = Form(...),
    projectDescription: str = Form(...),
    db: Session = Depends(database.get_db),
    current_user: models.User = Depends(get_current_user),
):
    """Upload a new dataset file (CSV, TSV, JSON, XLSX, or Parquet) for a project.

    Validates the file, stores it with a sanitized name, creates a working copy
    in the same native format, and returns the initial project data.
    """
    logger.info("Upload request: project=%s, file=%s", projectName, file.filename)
    await validate_upload_file(file)

    original_path, copy_path = store_upload(file)
    try:
        df = read_table_safe(original_path)
    except HTTPException as e:
        # The just-uploaded file could not be parsed — that's a bad client file,
        # not a server fault. Discard the orphaned files and report a clean 400.
        delete_project_files(str(copy_path))
        if e.status_code >= 500:
            ext = Path(file.filename).suffix.lower()
            raise HTTPException(status_code=400, detail=f"Could not parse the uploaded {ext} file.") from e
        raise

    project = create_project(db, projectName, str(copy_path), projectDescription, current_user.id)

    total_rows = len(df)
    resp = dataframe_to_response(df)
    return {
        "filename": project.name,
        "file_path": project.file_path,
        "project_id": project.project_id,
        "page": 1,
        "page_size": total_rows,
        "total_rows": total_rows,
        "total_pages": 1,
        **resp,
    }


@router.get("/get/{project_id}", response_model=schemas.ProjectResponse)
async def get_project_details(
    page: int = 1,
    pageSize: int = 50,
    project: models.Project = Depends(get_project_or_404),
):
    """Fetch full project details including all rows and columns."""
    df = read_table_safe(project.file_path)

    total_rows = len(df)
    total_pages = (total_rows + pageSize - 1) // pageSize

    start = (page - 1) * pageSize
    end = start + pageSize
    paginated_df = df.iloc[start:end]

    resp = dataframe_to_response(paginated_df)
    return {
        "filename": project.name,
        "file_path": project.file_path,
        "project_id": project.project_id,
        "page": page,
        "page_size": pageSize,
        "total_rows": total_rows,
        "total_pages": total_pages,
        **resp,
    }


@router.get("/recent", response_model=list[schemas.LastResponse])
def recent_projects(
    db: Session = Depends(database.get_db),
    current_user: models.User = Depends(get_current_user),
):
    """Get the current user's most recently modified projects."""
    projects = get_recent_projects(db, owner_id=current_user.id, limit=10)
    return [
        schemas.LastResponse(
            project_id=p.project_id,
            name=p.name,
            description=p.description,
            last_modified=p.last_modified,
        )
        for p in projects
    ]


@router.post("/{project_id}/save", response_model=schemas.ProjectResponse)
async def save_project(
    project_id: uuid.UUID,
    commit_message: str,
    db: Session = Depends(database.get_db),
    project: models.Project = Depends(get_project_or_404),
):
    """Save the current working dataset as a checkpoint.

    The working copy already reflects the user's latest accepted transforms, so
    checkpoint creation should preserve that current dataset and only update the
    checkpoint/log metadata for pending actions.
    """
    original_path = get_original_path(project.file_path)
    if Path(project.file_path).resolve() == original_path.resolve():
        logger.error(
            "Project working copy unexpectedly points at original file: id=%s working_copy=%s original=%s",
            project_id,
            project.file_path,
            original_path,
        )
        raise HTTPException(
            status_code=500,
            detail=f"Project {project_id} working copy is misconfigured; please retry or contact support.",
        )

    df = read_table_safe(project.file_path)

    # Create checkpoint (marks logs as applied)
    checkpoint = create_checkpoint(db, project_id, commit_message)

    total_rows = len(df)
    resp = dataframe_to_response(df)
    logger.info("Project saved: id=%s, checkpoint=%s", project_id, checkpoint.id)
    return {
        "filename": project.name,
        "file_path": str(project.file_path),
        "project_id": project.project_id,
        "page": 1,
        "page_size": total_rows,
        "total_rows": total_rows,
        "total_pages": 1,
        **resp,
    }


@router.post("/{project_id}/revert", response_model=schemas.ProjectResponse)
async def revert_to_checkpoint(
    project_id: uuid.UUID,
    checkpoint_id: uuid.UUID = None,
    db: Session = Depends(database.get_db),
    project: models.Project = Depends(get_project_or_404),
):
    """Revert project to its original state or to a specific checkpoint.

    When checkpoint_id is provided, replays only the logs up to and including
    that checkpoint onto the original file. When None, reverts to the original
    uploaded state.
    """
    original_path = get_original_path(project.file_path)
    df = read_table_safe(original_path)

    if checkpoint_id is not None:
        checkpoint = (
            db.query(models.Checkpoint)
            .filter(
                models.Checkpoint.id == checkpoint_id,
                models.Checkpoint.project_id == project_id,
            )
            .first()
        )
        if not checkpoint:
            raise HTTPException(status_code=404, detail="Checkpoint not found")

        # Find all checkpoint IDs created at or before the target checkpoint
        eligible_checkpoint_ids = [
            c.id
            for c in db.query(models.Checkpoint)
            .filter(
                models.Checkpoint.project_id == project_id,
                models.Checkpoint.created_at <= checkpoint.created_at,
            )
            .all()
        ]

        logs = (
            db.query(models.ProjectChangeLog)
            .filter(
                models.ProjectChangeLog.project_id == project_id,
                models.ProjectChangeLog.checkpoint_id.in_(eligible_checkpoint_ids),
                models.ProjectChangeLog.applied == True,  # noqa: E712
            )
            .order_by(models.ProjectChangeLog.timestamp)
            .all()
        )

        for log in logs:
            df = apply_logged_transformation(df, log.action_type, log.action_details)

    # Write file first — if this fails, DB is unchanged and state remains consistent.
    save_table_safe(df, project.file_path)
    # Clear unapplied logs so a subsequent save cannot re-apply stale
    # transformations on top of the reverted file state.
    # Applies to all reverts (full and partial) to prevent stale log replay.
    db.query(models.ProjectChangeLog).filter(
        models.ProjectChangeLog.project_id == project_id,
        models.ProjectChangeLog.applied.is_(False),
    ).delete(synchronize_session="evaluate")
    try:
        db.commit()
    except SQLAlchemyError:
        db.rollback()
        raise

    total_rows = len(df)
    resp = dataframe_to_response(df)
    logger.info("Project reverted: id=%s, checkpoint_id=%s", project_id, checkpoint_id)
    return {
        "filename": project.name,
        "file_path": project.file_path,
        "project_id": project.project_id,
        "page": 1,
        "page_size": total_rows,
        "total_rows": total_rows,
        "total_pages": 1,
        **resp,
    }


@router.get("/{project_id}/export")
async def export_project(project: models.Project = Depends(get_project_or_404)):
    """Download the current working copy of a project in its native format."""
    fmt = get_format(project.file_path)
    return FileResponse(
        project.file_path,
        media_type=fmt.media_type,
        filename=f"{project.name}{fmt.extension}",
    )


@router.delete("/{project_id}")
async def delete_project_endpoint(
    db: Session = Depends(database.get_db),
    project: models.Project = Depends(get_project_or_404),
):
    """Delete a project and its associated files."""
    delete_project_files(project.file_path)
    delete_project(db, project)
    return {"success": True, "message": "Project deleted"}


@router.post("/{project_id}/undo", response_model=schemas.ProjectResponse)
async def undo_last_transformation(
    project_id: uuid.UUID,
    db: Session = Depends(database.get_db),
):
    """Undo the most recent transformation.

    Removes the last change log entry and rebuilds the working copy
    by replaying all remaining logs onto the original file.
    """
    project = get_project_or_404(project_id, db)

    last_log = get_last_change_log(db, project_id)
    if not last_log:
        raise HTTPException(status_code=404, detail="No transformations to undo")

    delete_change_log(db, last_log)

    original_path = get_original_path(project.file_path)
    df = read_table_safe(original_path)

    remaining_logs = (
        db.query(models.ProjectChangeLog)
        .filter(models.ProjectChangeLog.project_id == project_id)
        .order_by(models.ProjectChangeLog.timestamp)
        .all()
    )

    for log in remaining_logs:
        df = apply_logged_transformation(df, log.action_type, log.action_details)

    save_table_safe(df, project.file_path)
    db.commit()

    resp = dataframe_to_response(df)
    logger.info(
        "Undo: project_id=%s, removed log_id=%s, remaining_logs=%d",
        project_id,
        last_log.change_log_id,
        len(remaining_logs),
    )
    return {
        "filename": project.name,
        "file_path": project.file_path,
        "project_id": project.project_id,
        "page": 1,
        "page_size": len(df),
        "total_rows": len(df),
        "total_pages": 1,
        **resp,
    }

File: dataloom-frontend/src/api/projects.ts
/**
 * API functions for project CRUD operations.
 * @module api/projects
 */
import client from "./client";
import type { Pagination, ProjectSummary, TableResponse } from "./types";

/** Backend `ProjectResponse` — full project payload with paginated rows. */
export interface ProjectDetails extends TableResponse, Pagination {
  filename: string;
  file_path: string;
  project_id: string;
}

/** Export options accepted by {@link exportProject}. */
export interface ExportOptions {
  /** Target format extension (e.g. "csv", "json"). */
  format?: string;
  /** CSV delimiter option: comma, tab, semicolon, or pipe. */
  delimiter?: string;
  /** Whether to include the header row. */
  includeHeader?: boolean;
  /** Output encoding: utf-8, latin-1, ascii, or utf-16. */
  encoding?: string;
}

/** The downloaded file plus the server-provided filename. */
export interface ExportResult {
  blob: Blob;
  filename: string | null;
}

/**
 * Upload a new project CSV file.
 * @param file - The CSV file to upload.
 * @param projectName - Name for the new project.
 * @param projectDescription - Description for the new project.
 * @returns The created project response.
 */
export const uploadProject = async (
  file: File,
  projectName: string,
  projectDescription: string,
): Promise<ProjectDetails> => {
  const formData = new FormData();
  formData.append("file", file);
  formData.append("projectName", projectName);
  formData.append("projectDescription", projectDescription);
  const response = await client.post("/projects/upload", formData);
  return response.data;
};

/**
 * Fetch full project details including rows and columns.
 * @param projectId - The project ID.
 * @param page - Current page.
 * @param pageSize - Elements per page.
 * @returns Project details with columns and rows.
 */
export const getProjectDetails = async (
  projectId: string,
  page?: number,
  pageSize?: number,
): Promise<ProjectDetails> => {
  const response = await client.get(`/projects/get/${projectId}`, {
    params: { page, pageSize },
  });
  return response.data;
};

/**
 * Fetch the most recently modified projects.
 * @returns List of recent project summaries.
 */
export const getRecentProjects = async (): Promise<ProjectSummary[]> => {
  const response = await client.get("/projects/recent");
  return response.data;
};

/**
 * Save the current project state as a checkpoint.
 * @param projectId - The project ID.
 * @param commitMessage - Description of changes.
 * @returns Updated project response.
 */
export const saveProject = async (projectId: string, commitMessage: string) => {
  const response = await client.post(
    `/projects/${projectId}/save?commit_message=${encodeURIComponent(commitMessage)}`,
  );
  return response.data;
};

/**
 * Revert project to a previous checkpoint.
 * @param projectId - The project ID.
 * @param checkpointId - The checkpoint ID to revert to.
 * @returns Reverted project response.
 */
export const revertToCheckpoint = async (projectId: string, checkpointId: string) => {
  const response = await client.post(`/projects/${projectId}/revert?checkpoint_id=${checkpointId}`);
  return response.data;
};

/**
 * Export a project's working copy, optionally converting to another format.
 * @param projectId - The project ID.
 * @param options - Target format extension or export options.
 * @returns The file blob and the server-provided download filename (parsed
 *   from Content-Disposition).
 */
export const exportProject = async (
  projectId: string,
  options?: string | ExportOptions,
): Promise<ExportResult> => {
  const exportOptions: ExportOptions =
    typeof options === "string" ? { format: options } : options || {};
  const { format, delimiter, includeHeader, encoding } = exportOptions;
  const params = {
    ...(format ? { format } : {}),
    ...(delimiter ? { delimiter } : {}),
    ...(includeHeader !== undefined ? { include_header: includeHeader } : {}),
    ...(encoding ? { encoding } : {}),
  };

  try {
    const response = await client.get(`/projects/${projectId}/export`, {
      params: Object.keys(params).length > 0 ? params : undefined,
      responseType: "blob",
    });
    const disposition: string = response.headers["content-disposition"] || "";
    const match = disposition.match(/filename="?([^"]+)"?/);
    return { blob: response.data, filename: match?.[1] ?? null };
  } catch (err) {
    // Under responseType:"blob" an error body arrives as a Blob, so the shared
    // interceptor can't read it. Surface the real backend detail (e.g. a 400
    // message) before re-throwing.
    const response = (err as { response?: { status?: number; data?: unknown } } | null)?.response;
    const data = response?.data;
    if (data instanceof Blob) {
      const detail = await data.text().catch(() => "");
      if (detail) console.error("Export failed:", response?.status, detail);
    }
    throw err;
  }
};

/**
 * Delete a project and its associated files.
 * @param projectId - The project ID.
 * @returns Success confirmation.
 */
export const deleteProject = async (projectId: string) => {
  const response = await client.delete(`/projects/${projectId}`);
  return response.data;
};

/**
 * Rename a project.
 * @param projectId - The project ID.
 * @param name - The new project name.
 * @returns Updated project response.
 */
export const renameProject = async (projectId: string, name: string) => {
  const response = await client.patch(`/projects/${projectId}/rename`, { name });
  return response.data;
};

/**
 * Search a project.
 * @param query - The search query.
 * @returns List of matched projects.
 */
export const searchProjects = async (query: string): Promise<ProjectSummary[]> => {
  const response = await client.get("/projects/search", { params: { q: query } });
  return response.data;
};

/**
 * Update project name and/or description.
 */
export const updateProject = async (
  projectId: string,
  { name, description }: { name?: string; description?: string } = {},
) => {
  const response = await client.patch(`/projects/${projectId}`, {
    ...(name !== undefined ? { name } : {}),
    ...(description !== undefined ? { description } : {}),
  });
  return response.data;
};

/**
 * Fetch project metadata only — no row data.
 * @param projectId - The project ID.
 * @returns Project metadata.
 */
export const getProjectMeta = async (projectId: string): Promise<ProjectSummary> => {
  const response = await client.get(`/projects/${projectId}/meta`);
  return response.data;
};

/**
 * Fetch a list of projects with optional pagination.
 * @param options - Pagination options.
 * @returns List of projects.
 */
export const getProjects = async ({
  limit = 50,
  offset = 0,
}: { limit?: number; offset?: number } = {}): Promise<ProjectSummary[]> => {
  const response = await client.get("/projects", {
    params: { limit, offset },
  });
  return response.data;
};
